File: androidApp/src/main/java/com/commuto/interfacemobile/android/p2p/PublicKeyAnnouncementParser.kt
package com.commuto.interfacemobile.android.p2p

import com.commuto.interfacemobile.android.key.keys.PublicKey
import com.commuto.interfacemobile.android.p2p.messages.PublicKeyAnnouncement
import com.commuto.interfacemobile.android.p2p.serializable.messages.SerializablePublicKeyAnnouncementMessage
import com.commuto.interfacemobile.android.p2p.serializable.payloads.SerializablePublicKeyAnnouncementPayload
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import java.nio.ByteBuffer
import java.nio.charset.Charset
import java.security.MessageDigest
import java.util.*

/**
 * Attempts to restore a [PublicKeyAnnouncement] from a given [String].
 *
 * @param messageString An optional [String] from which to try to restore a
 * [PublicKeyAnnouncement].
 *
 * @return A [PublicKeyAnnouncement] if [messageString] contains a valid
 * [Public Key Announcement](https://github.com/jimmyneutront/commuto-whitepaper/blob/main/commuto-interface-specification.txt),
 * or `null` if it does not.
 */
fun parsePublicKeyAnnouncement(messageString: String?): PublicKeyAnnouncement? {
    // Setup decoder
    val decoder = Base64.getDecoder()
    // Return null if no message string is given
    if (messageString == null) {
        return null
    }
    // Restore message object
    val message = try {
        Json.decodeFromString<SerializablePublicKeyAnnouncementMessage>(messageString)
    } catch (e: Exception) {
        return null
    }
    // Ensure that the message is a Public Key announcement message
    if (message.msgType != "pka") {
        return null
    }

    // Get the interface ID of the sender
    val senderInterfaceID = try {
        decoder.decode(message.sender)
    } catch (e: Exception) {
        return null
    }

    // Restore payload object
    val payloadBytes = try {
        decoder.decode(message.payload)
    } catch (e: Exception) {
        return null
    }
    val payload = try {
        val payloadString = payloadBytes.toString(Charset.forName("UTF-8"))
        Json.decodeFromString<SerializablePublicKeyAnnouncementPayload>(payloadString)
    } catch (e: Exception) {
        return null
    }

    // Get the offer ID in the announcement
    val messageOfferId = try {
        val offerIdByteArray = decoder.decode(payload.offerId)
        val offerIdByteBuffer = ByteBuffer.wrap(offerIdByteArray)
        val mostSigBits = offerIdByteBuffer.long
        val leastSigBits = offerIdByteBuffer.long
        UUID(mostSigBits, leastSigBits)
    } catch (e: Exception) {
        return null
    }

    // Re-create maker's public key
    val publicKey = try {
        PublicKey(decoder.decode(payload.pubKey))
    } catch (e: Exception) {
        return null
    }

    // Check that interface id of maker's key matches value in "sender" field of message
    if (!senderInterfaceID.contentEquals(publicKey.interfaceId)) {
        return null
    }

    // Create hash of payload
    val payloadDataHash = MessageDigest.getInstance("SHA-256").digest(payloadBytes)

    // Verify signature
    return try {
        when (publicKey.verifySignature(payloadDataHash, decoder.decode(message.signature))) {
            true -> PublicKeyAnnouncement(messageOfferId, publicKey)
            false -> null
        }
    } catch (e: Exception) {
        null
    }
}

File: iosApp/iosApp/p2p/New Group/PublicKeyAnnouncementParser.swift
//
//  PublicKeyAnnouncementParser.swift
//  iosApp
//
//  Created by jimmyt on 7/29/22.
//  Copyright © 2022 orgName. All rights reserved.
//

import CryptoKit
import Foundation

/**
Attempts to restore a PublicKeyAnnouncement from a given `String`.
 
 - Parameter messageString: An optional`String` from which to try to restore a `PublicKeyAnnouncement`.
 
 - Returns: An optional `PublicKeyAnnouncement` that will be `nil` if `messageString` does not contain a valid public key announcement, and will be non-`nil` if `messageString` does contain a valid public key announcement.
 */
func parsePublicKeyAnnouncement(messageString: String?) -> PublicKeyAnnouncement? {
    guard messageString != nil else {
        return nil
    }
    // Restore message NSDictionary
    guard let messageData = messageString!.data(using: String.Encoding.utf8) else {
        return nil
    }
    guard let message = try? JSONSerialization.jsonObject(with: messageData) as? NSDictionary else {
        return nil
    }
    
    // Ensure that the message is a Public Key Announcement message
    guard let messageType = message["msgType"] as? String else {
        return nil
    }
    guard messageType == "pka" else {
        return nil
    }
    
    // Get the interface ID of the sender
    guard let senderInterfaceIdString = message["sender"] as? String, let senderInterfaceId = Data(base64Encoded: senderInterfaceIdString) else {
        return nil
    }
    
    // Restore payload NSDictionary
    guard let payloadString = message["payload"] as? String, let payloadData = Data(base64Encoded: payloadString) else {
        return nil
    }
    guard let payload = try? JSONSerialization.jsonObject(with: payloadData) as? NSDictionary else {
        return nil
    }
    
    // Get the offer ID in the announcement
    guard let messageOfferIdString = payload["offerId"] as? String, let messageOfferIdData = Data(base64Encoded: messageOfferIdString), let messageOfferId = UUID.from(data: messageOfferIdData) else {
        return nil
    }
    
    // Re-create maker's public key
    guard let pubKeyString = payload["pubKey"] as? String, let pubKeyBytes = Data(base64Encoded: pubKeyString) else {
        return nil
    }
    guard let publicKey = try? PublicKey(publicKeyBytes: pubKeyBytes) else {
        return nil
    }
    
    // Check that interface id of maker's key matches value in "sender" field of message
    guard senderInterfaceId == publicKey.interfaceId else {
        return nil
    }
    
    // Create hash of payload
    let payloadDataDigest = SHA256.hash(data: payloadData)
    var payloadDataHashByteArray = [UInt8]()
    for byte: UInt8 in payloadDataDigest.makeIterator() {
        payloadDataHashByteArray.append(byte)
    }
    let payloadDataHash = Data(bytes: payloadDataHashByteArray, count: payloadDataHashByteArray.count)
    
    // Verify signature
    guard let signatureString = message["signature"] as? String, let signature = Data(base64Encoded: signatureString) else {
        return nil
    }
    if (try? publicKey.verifySignature(signedData: payloadDataHash, signature: signature)) != nil {
        return PublicKeyAnnouncement(offerId: messageOfferId, pubKey: publicKey)
    } else {
        return nil
    }
}
